crypto/asn1/a_digest.c

```diff
@@ -48,20 +48,35 @@ int ASN1_digest(i2d_of_void *i2d, const EVP_MD *type, char *data,
 
 #endif
 
-int ASN1_item_digest(const ASN1_ITEM *it, const EVP_MD *type, void *asn,
-                     unsigned char *md, unsigned int *len)
+int ASN1_item_digest_with_libctx(const ASN1_ITEM *it, const EVP_MD *md,
+                                 void *asn, unsigned char *data,
+                                 unsigned int *len, OPENSSL_CTX *libctx,
+                                 const char *propq)
 {
-    int i;
+    int i, ret = 0;
     unsigned char *str = NULL;
+    EVP_MD *fetched_md = (EVP_MD *)md;
 
     i = ASN1_item_i2d(asn, &str, it);
-    if (!str)
+    if (str == NULL)
         return 0;
 
-    if (!EVP_Digest(str, i, md, len, type, NULL)) {
-        OPENSSL_free(str);
-        return 0;
-    }
+    if (EVP_MD_provider(md) == NULL)
+         fetched_md = EVP_MD_fetch(libctx, EVP_MD_name(md), propq);
```

Comment:
What if `md` is an ENGINE provided implementation? Should we prioritise the ENGINE implementation over a fetched one? In some cases (e.g. GOST??) there might not even be a fetched implementation.

Reply:
What are you suggesting this code should look like? Do you mean this??
     if (fetched_md == NULL)
         fetched_md = md;

Reply:
Perhaps this?:

````
if (EVP_MD_provider(md) == NULL && ENGINE_get_digest_engine(EVP_MD_type(md)) == NULL)
    fetched_md = EVP_MD_fetch(libctx, EVP_MD_name(md), propq);
````

Reply:
Ok I have changed it..
I would imagine this is not going to be the only case where this is going to happen.